File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        pref = [0] * len(nums)
        pref[0] = nums[0]
        for i in range(1, len(nums)):
            pref[i] = pref[i-1] + nums[i]
        
        m = [0] * k
        m[0] = 1
        ans = 0
        for x in pref:
            if x < 0:
                x = (x % k + k) % k
            m[x % k] += 1
        # n choose 2
        for x in m:
            ans += x * (x - 1) // 2
        return ans

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    vector<int> fun(vector<int>& a) {
        int n = a.size();
        vector<int> pref(n);
        pref[0] = a[0];
        for (int i = 1; i < n; i++) pref[i] = pref[i - 1] + a[i];
        return pref;
    }

    int subarraysDivByK(vector<int>& nums, int k) {
        vector<int> pref = fun(nums);
        vector<int> m(k);
        m[0] = 1;
        int ans = 0;
        for (auto& x : pref) {
            // if x is negative, turn it to positive modular equivalent
            if (x < 0) x = (x % k + k) % k;m[x % k]++;
        }
        for (auto& x : m) ans += x * (x - 1) / 2;
        return ans;
    }
};